nitrite.go

```diff
@@ -1,201 +1,49 @@
-	ExternalAAD []byte
-	Payload     []byte
-}
-
-// Size of these fields (in bytes) comes from AWS Nitro documentation at
-// https://docs.aws.amazon.com/enclaves/latest/user/enclaves-user.pdf
-// from May 4, 2022.
-// An experiment on 08/22/24 shows the following configuration is also possible
-//
-//	MaxUserDataLen  = 4050
-//	MaxNonceLen     = 0
-//	MaxPublicKeyLen = 0
-const (
-	MaxUserDataLen  = 1024
-	MaxNonceLen     = MaxUserDataLen
-	MaxPublicKeyLen = MaxUserDataLen
-)
-
-// Errors that are encountered when manipulating the COSESign1 structure.
-var (
-	ErrBadCOSESign1Structure          error = errors.New("Data is not a COSESign1 array")
-	ErrCOSESign1EmptyProtectedSection error = errors.New("COSESign1 protected section is nil or empty")
-	ErrCOSESign1EmptyPayloadSection   error = errors.New("COSESign1 payload section is nil or empty")
-	ErrCOSESign1EmptySignatureSection error = errors.New("COSESign1 signature section is nil or empty")
-	ErrCOSESign1BadAlgorithm          error = errors.New("COSESign1 algorithm not ECDSA384")
-)
-
-// Errors encountered when parsing the CoseBytes attestation document.
-var (
-	ErrBadAttestationDocument error = errors.New("Bad attestation document")
-	ErrMandatoryFieldsMissing error = errors.New("One or more of mandatory fields missing")
-	ErrBadDigest              error = errors.New("Payload 'digest' is not SHA384")
-	ErrBadTimestamp           error = errors.New("Payload 'timestamp' is 0 or less")
-	ErrBadPCRs                error = errors.New("Payload 'pcrs' is less than 1 or more than 32")
-	ErrBadPCRIndex            error = errors.New("Payload 'pcrs' key index is not in [0, 32)")
-	ErrBadPCRValue            error = errors.New("Payload 'pcrs' value is nil or not of length {32,48,64}")
-	ErrBadCABundle            error = errors.New("Payload 'cabundle' has 0 elements")
-	ErrBadCABundleItem        error = errors.New("Payload 'cabundle' has a nil item or of length not in [1, 1024]")
-	ErrBadPublicKey           error = fmt.Errorf(
-		"Payload 'public_key' length greater than %d",
-		MaxPublicKeyLen,
-	)
-	ErrBadUserData error = fmt.Errorf(
-		"Payload 'user_data' length greater than %d",
-		MaxUserDataLen,
-	)
-	ErrBadNonce error = fmt.Errorf(
-		"Payload 'nonce' length greater than %d",
-		MaxNonceLen,
-	)
-	ErrBadCertificatePublicKeyAlgorithm error = errors.New("Payload 'certificate' has a bad public key algorithm (not ECDSA)")
-	ErrBadCertificateSigningAlgorithm   error = errors.New("Payload 'certificate' has a bad public key signing algorithm (not ECDSAWithSHA384)")
-	ErrBadSignature                     error = errors.New("Payload's signature does not match signature from certificate")
-	ErrMarshallingCoseSignature         error = errors.New("Could not marshal COSE signature")
-)
-
-type VerificationTimeFunc func(Document) time.Time
-
-func WithTime(t time.Time) VerificationTimeFunc {
-	return func(_ Document) time.Time {
-		return t
+func NewVerifier(
+	certProviderOpt func() (CertProvider, error),
```

Comment:
I want this option to be mandatory - on has to select a cert provider.

If we like this approach, I'll also add one for self-signed certs so that that functionality doesn't get lost.